test/client-unit/services/PurchaseOrderFormService.spec.js

```diff
@@ -0,0 +1,148 @@
+    _$timeout_,
+  ) => {
+
+    PurchaseOrderForm = _PurchaseOrderForm_;
+    Mocks = _MockDataService_;
+    Session = _SessionService_;
+    $timeout = _$timeout_;
+
+    // set up the required properties for the session
+    Session.create(Mocks.user(), Mocks.enterprise(), Mocks.project());
+
+    httpBackend = $httpBackend;
+
+    httpBackend.when('GET', '/inventory/metadata/?locked=0&use_previous_price=1')
+      .respond(200, Mocks.inventories());
+
+  }));
+
+  beforeEach(() => {
+    form = new PurchaseOrderForm('TestKey');
+
+    // Load the inventory
+    httpBackend.flush();
+  });
+
+  // make sure $http is clean after tests
+  afterEach(() => {
+    $timeout.flush();
+    httpBackend.verifyNoOutstandingExpectation();
+    httpBackend.verifyNoOutstandingRequest();
+  });
+
+  describe('calls #setup() in the constructor()', () => {
+    let setupSpy;
+
+    before(() => {
+      setupSpy = chai.spy.on(form, 'setup');
+    });
+
+    it('calls #setup() on initialization', () => {
+      expect(setupSpy).to.have.been.called;
+    });
+
+    it('sets up store.data as an empty array', () => {
+      expect(form.store.data).to.be.a('array');
+      expect(form.store.data).to.have.length(0);
+    });
+  });
+
+  it('produces an error when #validate() is called with bad PO items', () => {
+    // Add a bad (uninitialized) PO item
+    form.addItem();
+
+    // Validate the form (should fail)
+    form.validate();
+    expect(form._valid).to.be.false;
+    expect(form._invalid).to.be.true;
+  });
+
+  it('adds an empty item when addItem() is called ', () => {
+    // There should be no items before addItem is called
+    expect(form.store.data.length).to.equal(0);
+
+    // Add an item and verify that one got added
+    form.addItem();
+    expect(form.store.data.length).to.equal(1);
+  });
+
+  it('deletes an item when when #removeItem() is callled', () => {
+    // Make sure it is empty to start with
+    expect(form.store.data.length).to.equal(0);
+
+    // Create a Purchase Order Item on the form
+    const item = form.addItem();
+    item.inventory_uuid = form.inventory.available.data[0].uuid;
+    form.configureItem(item);
+
+    // Now that we have added the item, the count should be 1
+    expect(form.store.data.length).to.equal(1);
+
+    // Remove the item and make sure the count goes back to 0
+    form.removeItem(item);
+    expect(form.store.data.length).to.equal(0);
+  });
+
+  it('sets totals correctly when #digest() is callled', () => {
+
+    // Get the UUIDs for a couple of inventory items
+    const uuid1 = form.inventory.available.data[0].uuid;
+    const uuid2 = form.inventory.available.data[1].uuid;
+
+    // First, verify that the initial cost is 0
+    expect(form.details.cost).to.equal(0);
+
+    // Add a PO item to the form
+    const item1 = form.addItem();
+    item1.inventory_uuid = uuid1;
+    form.configureItem(item1);
+    item1.quantity = 2;
+    item1.unit_price = 2.30;
+    item1._hasValidAccounts = true;
```

Comment:
You are implicitly testing both the `PurchaseOrderForm` and the `PurchaseOrderFormItem` here.  Each item is actually a subclass to itself.  When you call `configure()` on it, it should validate that it has all the correct accounts set.

I think this is fine, even if a bit clunky.  The only way to make this better would be to improve the mock data .. but a better order of operations would be to write unit tests for the `PurchaseOrderItem` service (thereby forcing the development of that mock data) and then use the mock data from that to power the "correct" scenario here.

Reply:
@jniles Thanks for looking at these!